frontend/src/test/scala/bloop/bsp/BspProtocolSpec.scala

```diff
@@ -344,6 +345,76 @@ class BspProtocolSpec(
     }
   }
 
+  test("sources globs are expanded in sources request") {
+    TestUtil.withinWorkspace { workspace =>
+      val logger = new RecordingLogger(ansiCodesSupported = false)
+      object Sources {
+        val `Hello.scala` =
+          """/Hello.scala
+            |object A
+          """.stripMargin
+        val `HelloTest.scala` =
+          """/HelloTest.scala
+            |object HelloTest
+          """.stripMargin
+
+        // This file is ignored because `walkDepth == 1`, meaning that the globs
+        // only expand to file that are depth 1 from the base directory.
+        val `toodeep/Hello.scala` =
+          """/toodeep/HelloTest.scala
+            |package toodeep
+            |object Hello
+          """.stripMargin
+      }
+
+      val globDirectory =
+        workspace.underlying.resolve("a").resolve("src")
+      val baseProject = TestProject(
+        workspace,
+        "a",
+        sources = List(
+          Sources.`Hello.scala`,
+          Sources.`HelloTest.scala`,
+          Sources.`toodeep/Hello.scala`
+        ),
+        sourcesGlobs = List(
+          Config.SourcesGlobs(
+            globDirectory,
+            walkDepth = Some(1),
+            includes = List("*.scala"),
+            excludes = List("*Test.scala")
+          )
+        )
+      )
+      val `A` = baseProject.copy(config = baseProject.config.copy(sources = Nil))
+
+      val projects = List(`A`)
+      loadBspState(workspace, projects, logger) { state =>
+        val project = state.underlying.build.loadedProjects.head
+        def assertSourcesMatches(expected: String): Unit = {
+          val obtained = for {
+            item <- state.requestSources(`A`).items
+            source <- item.sources
+          } yield globDirectory.relativize(source.uri.toPath).iterator().asScala.mkString("/")
```

Comment:
In `scala.meta.io.RelativePath.toRelative(isDirectory: Boolean)` we do the same processing: `.iterator().asScala.mkString("/")` 

Worth adding to relative path? `nio.Path.toUri` absolutizes relative paths

Reply:
Yes, I think it makes sense to do so :+1:

Reply:
Sounds good. Done, I ported the implementation from `scala.meta.io.RelativePath.toUri()` which handles trickier cases like URI-encoding individual parts of paths.